hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -163,66 +150,64 @@ public ReferenceCounted<IOmMetadataReader, SnapshotCache> get(String key)
-              // Unexpected and unknown exception thrown from CacheLoader#load
-              throw new IllegalStateException(ex);
             }
-          }
-          return v;
-        });
-
-    if (rcOmSnapshot == null) {
-      // The only exception that would fall through the loader logic above
-      // is OMException with FILE_NOT_FOUND.
-      throw new OMException("Snapshot table key '" + key + "' not found, "
-          + "or the snapshot is no longer active",
-          OMException.ResultCodes.FILE_NOT_FOUND);
-    } else {
-      // When RC OmSnapshot is successfully loaded
-      rcOmSnapshot.incrementRefCount();
-    }
-
-    // If the snapshot is already loaded in cache, the check inside the loader
-    // above is ignored. But we would still want to reject all get()s except
-    // when called from SDT (and some) if the snapshot is not active anymore.
-    if (!skipActiveCheck &&
-        !omSnapshotManager.isSnapshotStatus(key, SNAPSHOT_ACTIVE)) {
-      // Ref count was incremented. Need to decrement on exception here.
-      rcOmSnapshot.decrementRefCount();
-      throw new OMException("Unable to load snapshot. " +
-          "Snapshot with table key '" + key + "' is no longer active",
-          FILE_NOT_FOUND);
-    }
+            if (v != null) {
+              // When RC OmSnapshot is successfully loaded
+              v.incrementRefCount();
+            }
+            return v;
+          });
+      if (rcOmSnapshot == null) {
+        // The only exception that would fall through the loader logic above
+        // is OMException with FILE_NOT_FOUND.
+        throw new OMException("Snapshot table key '" + key + "' not found, "
+            + "or the snapshot is no longer active",
+            OMException.ResultCodes.FILE_NOT_FOUND);
+      }
 
-    synchronized (pendingEvictionList) {
-      // Remove instance from clean up list when it exists.
-      pendingEvictionList.remove(rcOmSnapshot);
+      // If the snapshot is already loaded in cache, the check inside the loader
+      // above is ignored. But we would still want to reject all get()s except
+      // when called from SDT (and some) if the snapshot is not active anymore.
+      if (!skipActiveCheck &&
+          !omSnapshotManager.isSnapshotStatus(key, SNAPSHOT_ACTIVE)) {
+        // Ref count was incremented. Need to decrement on exception here.
+        rcOmSnapshot.decrementRefCount();
+        throw new OMException("Unable to load snapshot. " +
+            "Snapshot with table key '" + key + "' is no longer active",
+            FILE_NOT_FOUND);
+      }
+    } finally {
+      lock.unlock();
     }
 
     // Check if any entries can be cleaned up.
     // At this point, cache size might temporarily exceed cacheSizeLimit
-    // even if there are entries that can be evicted, which is fine since it
-    // is a soft limit.
+    // even if there are entries that can be evicted, which is fine since it is a soft limit.
     cleanup();
```

Comment:
Now that this is moved outside of the lock I think it should be fine with the deadlock case.

Reply:
Yes, it should be fine as explained in https://github.com/apache/ozone/pull/5968#issuecomment-1885986128